File: android/src/main/java/com/bixolon/bxlflutterbgatelib/BxlMPosControllerConfigPlugin.java
package com.bixolon.bxlflutterbgatelib;
import android.content.Context;
import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import com.bixolon.mpos.MPosControllerConfig;
import com.bixolon.commonlib.BXLCommonConst;
import com.bixolon.commonlib.log.LogService;
import org.json.JSONArray;
import org.json.JSONObject;

import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.Semaphore;
import java.nio.ByteBuffer;

/** BxlMPosControllerConfigPlugin */
public class BxlMPosControllerConfigPlugin implements FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private MethodChannel channel;
    private MPosControllerConfig config = new MPosControllerConfig();
    private Context context;
    private final Semaphore available = new Semaphore(1, true);

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        final String METHOD_CHANNEL = "com.bixolon.mposcontroller/config";
        BinaryMessenger messenger = flutterPluginBinding.getBinaryMessenger();
        context = flutterPluginBinding.getApplicationContext();
        channel = new MethodChannel(messenger, METHOD_CHANNEL);
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        class NewRunnable implements Runnable {
            @Override
            public void run() {
                try {
                    available.acquire();
                    switch (call.method)
                    {
                        case "selectInterface":             selectInterface(call, result); break;
                        case "selectCommandMode":           selectCommandMode(call, result); break;
                        case "openService":                 openService(call, result); break;
                        case "closeService":                closeService(call, result); break;
                        case "isOpen":                      isOpen(call, result); break;
                        case "getDeviceId":                 getDeviceId(call, result); break;
                        case "directIO":                    directIO(call, result); break;
                        case "setReadMode":                 setReadMode(call, result); break;
                        case "setKeepNetworkConnection":    setKeepNetworkConnection(call, result); break;
                        case "setTransaction":              setTransaction(call, result); break;
                        case "searchDevices":               searchDevices(call, result); break;
                        case "getBgateSerialNumber":        getBgateSerialNumber(call, result); break;
                        case "getUSBDevice":                getUSBDevice(call, result); break;
                        case "getCustomDevices":            getCustomDevices(call, result); break;
                        case "getSerialConfig":             getSerialConfig(call, result); break;
                        case "getSerialConfiguration":      getSerialConfiguration(call, result); break;
                        case "setSerialConfig":             setSerialConfig(call, result); break;
                        case "setSerialConfiguration":      setSerialConfiguration(call, result); break;
                        case "addCustomDevice":             addCustomDevice(call, result); break;
                        case "deleteCustomDevice":          deleteCustomDevice(call, result); break;
                        case "reInitCustomDeviceType":      reInitCustomDeviceType(call, result); break;
                        default:
                            result.notImplemented();
                            break;
                    }
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }finally {
                    available.release();
                }
            }
        }
        NewRunnable nr = new NewRunnable();
        Thread t = new Thread(nr);
        t.start();
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }

    private Object getValue(@NonNull MethodCall call, @NonNull String key, @NonNull Object defaultValue) {
        return call.argument(key) == null ? defaultValue : call.argument(key);
    }

    public void selectInterface(@NonNull MethodCall call, @NonNull Result result) {
        Integer interfaceType = (Integer)getValue(call, "interface_type", -1);
        String address = (String)getValue(call, "address", "");
        try{
            result.success((int)config.selectInterface(interfaceType, address));
        }catch(Exception e){
            result.success(1000);
        }
    }

    public void selectCommandMode(@NonNull MethodCall call, @NonNull Result result) {
        Integer mode = (Integer)getValue(call, "command_mode", -1);
        try{
            result.success((int)config.selectCommandMode(mode));
        }catch(Exception e){
            result.success(1000);
        }
    }

    public void openService(@NonNull MethodCall call, @NonNull Result result) {
        Integer id = (Integer)getValue(call, "device_id", -1);
        Integer timeout = (Integer)getValue(call, "time_out", 3);
        try{
            int apiResult = (int)config.openService(id, timeout, context);
            result.success(apiResult);
        }catch(Exception e){
            result.success(1000);
        }
    }

    public void closeService(@NonNull MethodCall call, @NonNull Result result) {
        Integer timeout = (Integer)getValue(call, "time_out", 3);
        try{
            result.success((int)config.closeService(timeout));
        }catch(Exception e){
            result.success(1000);
        }
    }

    public void isOpen(@NonNull MethodCall call, @NonNull Result result) {
        try{
            result.success(config.isOpen() ? 1 : 0);
        }catch(Exception e){
            result.success(0);
        }
    }

    public void getDeviceId(@NonNull MethodCall call, @NonNull Result result) {
        try{
            result.success((int)config.getDeviceId());
        }catch(Exception e){
            result.success(-1);
        }
    }

    public void directIO(@NonNull MethodCall call, @NonNull Result result) {
        ArrayList<Integer> dataArray = call.argument("data_array") == null ? null : call.argument("data_array");
        if(dataArray == null){
            result.success(1000);
            return;
        }
        try{
            int index = 0;
            byte[] byteArray = new byte[dataArray.toArray().length];
            Iterator<Integer> iterator = dataArray.iterator();
            while(iterator.hasNext()){
                byteArray[index++] = iterator.next().byteValue();
            }
            result.success((int)config.directIO(byteArray));
        }catch(Exception e){
            result.success(1000);
        }
    }

    public void setReadMode(@NonNull MethodCall call, @NonNull Result result) {
        Integer mode = (Integer)getValue(call, "read_mode", -1);
        try{
            result.success((int)config.setReadMode(mode));
        }catch(Exception e){
            result.success(1000);
        }
    }

    public void setKeepNetworkConnection(@NonNull MethodCall call, @NonNull Result result) {
        Integer keep_connection = (Integer)getValue(call, "keep_connection", 0);
        result.success(1000);
    }

    public void setTransaction(@NonNull MethodCall call, @NonNull Result result) {
        Integer transaction = (Integer)getValue(call, "transaction", -1);
        try{
            result.success((int)config.setTransaction(transaction));
        }catch(Exception e){
            result.success(1000);
        }
    }

    public void searchDevices(@NonNull MethodCall call, @NonNull Result result) {
        JSONArray jsonArray = new JSONArray();
        try{
            int[] devices = config.searchDevices();
            if(devices == null){
                result.success(jsonArray.toString());
                return;
            }
            for(int index = 0; index < devices.length; index++){
                JSONObject jsonItem = new JSONObject();
                jsonItem.put("device_id", (int)devices[index]);
                if(devices[index] > 0 && devices[index] < 10)           jsonItem.put("device_type", "LABEL PRINTER");
                else if(devices[index] >= 10 && devices[index] < 20)    jsonItem.put("device_type", "POS PRINTER");
                else if(devices[index] >= 20 && devices[index] < 30)    jsonItem.put("device_type", "Unregistered HID INPUT DEVICE");
                else if(devices[index] >= 30 && devices[index] < 40)    jsonItem.put("device_type", "MSR");
                else if(devices[index] >= 40 && devices[index] < 50)    jsonItem.put("device_type", "BARCODE SCANNER");
                else if(devices[index] >= 60 && devices[index] < 70)    jsonItem.put("device_type", "RFID");
                else if(devices[index] >= 70 && devices[index] < 80)    jsonItem.put("device_type", "DALLAS READER");
                else if(devices[index] >= 80 && devices[index] < 90)    jsonItem.put("device_type", "NFC");
                else if(devices[index] >= 100 && devices[index] < 110)  jsonItem.put("device_type", "Unregistered USB-SERIAL");
                else if(devices[index] >= 110 && devices[index] < 120)  jsonItem.put("device_type", "CUSTOMER DISPLAY");
                else if(devices[index] >= 120 && devices[index] < 130)  jsonItem.put("device_type", "USB-SERIAL");
                else if(devices[index] >= 130 && devices[index] < 140)  jsonItem.put("device_type", "SCALE");
                else                                                    jsonItem.put("device_type", "UNKNOWN");
                // add json string
                jsonArray.put(jsonItem);
            }
            result.success(jsonArray.toString());
        }catch(Exception e){
            result.success(jsonArray.toString());
        }

    }

    public void getBgateSerialNumber(@NonNull MethodCall call, @NonNull Result result) {
        try{
            byte[] bytes = config.getBGateSerialNumber();
            if(bytes != null){
                result.success(new String(bytes, Charset.forName("UTF-8")));
            }else{
                result.success("");
            }
        }catch(Exception e){
            result.success("");
        }
    }

    public void getUSBDevice(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_id = (Integer)getValue(call, "device_id", -1);
        try{
            String resultString = config.getUSBDevice(device_id);
            //byte[] test = resultString.getBytes(Charset.forName("UTF-8"));
            result.success(resultString);
        }catch(Exception e){
            result.success("");
        }
    }

    public void getCustomDevices(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_type = (Integer)getValue(call, "device_type", -1);
        try{
            String resultString = config.getCustomDeviceList(device_type);
            result.success(resultString);
        }catch(Exception e){
            result.success(new ArrayList<String>());
        }
    }

    public void getSerialConfig(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_id = (Integer)getValue(call, "device_id", -1);
        try{
            byte[] configs = config.getSerialConfiguration(device_id);
            if(configs == null){
                result.success(new ArrayList<Integer>());
                return;
            }
            ArrayList<Integer> list = new ArrayList<Integer>();
            for(int index = 0; index < configs.length; index++){
                list.add((int) configs[index]);
            }
            result.success(list);
        }catch(Exception e){
            result.success(new ArrayList<Integer>());
        }
    }

    public void getSerialConfiguration(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_id = (Integer)getValue(call, "device_id", -1);
        try{
            byte[] configs = config.getSerialConfiguration(device_id);
            if(configs == null){
                result.success(new ArrayList<Integer>());
                return;
            }
            ArrayList<Integer> list = new ArrayList<Integer>();
            for(int index = 0; index < configs.length; index++){
                list.add((int) configs[index]);
            }
            result.success(list);
        }catch(Exception e){
            result.success(new ArrayList<Integer>());
        }
    }

    public void setSerialConfig(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_id = (Integer)getValue(call, "device_id", -1);
        Integer baud_rate = (Integer)getValue(call, "baud_rate", 2);
        Integer data_bit = (Integer)getValue(call, "data_bit", 1);
        Integer stop_bit = (Integer)getValue(call, "stop_bit", 0);
        Integer parity_bit = (Integer)getValue(call, "parity_bit", 0);
        Integer flow_control = (Integer)getValue(call, "flow_control", 1);
        try{
            long res = config.setSerialConfiguration(device_id,
                    baud_rate.byteValue(),
                    data_bit.byteValue(),
                    stop_bit.byteValue(),
                    parity_bit.byteValue(),
                    flow_control.byteValue());
            result.success(res);
        }catch (Exception e){
            result.success(1000);
        }
    }

    public void setSerialConfiguration(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_id = (Integer)getValue(call, "device_id", -1);
        Integer baud_rate = (Integer)getValue(call, "baud_rate", 2);
        Integer data_bit = (Integer)getValue(call, "data_bit", 1);
        Integer stop_bit = (Integer)getValue(call, "stop_bit", 0);
        Integer parity_bit = (Integer)getValue(call, "parity_bit", 0);
        Integer flow_control = (Integer)getValue(call, "flow_control", 1);
        try{
            long res = config.setSerialConfiguration(device_id,
                    baud_rate.byteValue(),
                    data_bit.byteValue(),
                    stop_bit.byteValue(),
                    parity_bit.byteValue(),
                    flow_control.byteValue());
            result.success(res);
        }catch (Exception e){
            result.success(1000);
        }
    }

    public void addCustomDevice(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_type = (Integer)getValue(call, "device_type", -1);
        String strVID = (String) getValue(call, "vid", "");
        String strPID = (String) getValue(call, "pid", "");
        try{
            ByteBuffer deviceBuf = ByteBuffer.allocate(4);
            deviceBuf.put(new java.math.BigInteger(strVID.substring(0, 2), 16).toByteArray());
            deviceBuf.put(new java.math.BigInteger(strVID.substring(2, 4), 16).toByteArray());
            deviceBuf.put(new java.math.BigInteger(strPID.substring(0, 2), 16).toByteArray());
            deviceBuf.put(new java.math.BigInteger(strPID.substring(2, 4), 16).toByteArray());
            result.success(config.addCustomDevice(device_type, deviceBuf.get(0), deviceBuf.get(1), deviceBuf.get(2), deviceBuf.get(3)));
        }catch(Exception e){
            result.success("");
        }
    }

    public void deleteCustomDevice(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_type = (Integer)getValue(call, "device_type", -1);
        String strVID = (String) getValue(call, "vid", "");
        String strPID = (String) getValue(call, "pid", "");
        try{
            ByteBuffer deviceBuf = ByteBuffer.allocate(4);
            deviceBuf.put(new java.math.BigInteger(strVID.substring(0, 2), 16).toByteArray());
            deviceBuf.put(new java.math.BigInteger(strVID.substring(2, 4), 16).toByteArray());
            deviceBuf.put(new java.math.BigInteger(strPID.substring(0, 2), 16).toByteArray());
            deviceBuf.put(new java.math.BigInteger(strPID.substring(2, 4), 16).toByteArray());
            result.success(config.deleteCustomDevice(device_type, deviceBuf.get(0), deviceBuf.get(1), deviceBuf.get(2), deviceBuf.get(3)));
        }catch(Exception e){
            result.success("");
        }
    }

    public void reInitCustomDeviceType(@NonNull MethodCall call, @NonNull Result result) {
        Integer device_type = (Integer)getValue(call, "device_type", -1);
        try{
            result.success(config.resetCustomDeviceList(device_type));
        }catch(Exception e){
            result.success("");
        }
    }
}

File: ios/Classes/BxlMPosControllerConfigPlugin.m
#import "BxlMPosControllerConfigPlugin.h"

@import frameworkMPosSDK;

static NSString* const PLATFORM_CHANNEL = @"com.bixolon.mposcontroller/config";

@interface BxlMPosControllerConfigPlugin()
@property MPosControllerConfig* config;
@end

@implementation BxlMPosControllerConfigPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName: PLATFORM_CHANNEL
                                     binaryMessenger:[registrar messenger]];
    
    BxlMPosControllerConfigPlugin* instance = [[BxlMPosControllerConfigPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (id)init{
    if ((self = [super init])) {
        self.config = [MPosControllerConfig new];
    }
    
    [MPosControllerDevices setEnableLog:YES saveFile:NO saveToHex:NO];
    
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    // Common APIs
    if ([@"openService" isEqualToString:call.method])
        return [self openService:call result: result];
    
    else if ([@"closeService" isEqualToString:call.method])
        return [self closeService:call result: result];
    
    else if ([@"selectInterface" isEqualToString:call.method])
        return [self selectInterface:call result: result];
    
    else if ([@"selectCommandMode" isEqualToString:call.method])
        return [self selectCommandMode:call result: result];
    
    else if ([@"setTransaction" isEqualToString:call.method])
        return [self setTransaction:call result: result];
    
    else if ([@"directIO" isEqualToString:call.method])
        return [self directIO:call result: result];
    
    else if ([@"isOpen" isEqualToString:call.method])
        return [self isOpen:call result: result];
    
    else if ([@"getDeviceId" isEqualToString:call.method])
        return [self getDeviceId:call result: result];
    
    else if ([@"setReadMode" isEqualToString:call.method])
        return [self setReadMode:call result: result];
    
    else if ([@"setKeepNetworkConnection" isEqualToString:call.method])
        return [self setKeepNetworkConnection:call result: result];
    
    else if ([@"isUsbPeripheralPrinter" isEqualToString:call.method])
        return [self isUsbPeripheralPrinter:call result:result];
    
    // Config APIs
    else if ([@"searchDevices" isEqualToString:call.method])
        return [self searchDevices:call result: result];
    
    else if ([@"getBgateSerialNumber" isEqualToString:call.method])
        return [self getBgateSerialNumber:call result: result];
    
    else if ([@"getUSBDevice" isEqualToString:call.method])
        return [self getUSBDevice:call result: result];
    
    else if ([@"getCustomDevices" isEqualToString:call.method])
        return [self getCustomDevices:call result: result];
    
    else if ([@"getSerialConfig" isEqualToString:call.method])
        return [self getSerialConfig:call result: result];
    
    else if ([@"getSerialConfiguration" isEqualToString:call.method])
        return [self getSerialConfiguration:call result: result];
    
    else if ([@"setSerialConfig" isEqualToString:call.method])
        return [self setSerialConfig:call result: result];
    
    else if ([@"setSerialConfiguration" isEqualToString:call.method])
        return [self setSerialConfiguration:call result: result];
    
    else if ([@"addCustomDevice" isEqualToString:call.method])
        return [self addCustomDevice:call result: result];
    
    else if ([@"deleteCustomDevice" isEqualToString:call.method])
        return [self deleteCustomDevice:call result: result];
    
    else if ([@"reInitCustomDeviceType" isEqualToString:call.method])
        return [self reInitCustomDeviceType:call result: result];
    
    else
        result(FlutterMethodNotImplemented);
}

// MARK: - Devices APIs

- (void)openService:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* deviceId = arguments[@"device_id"] ? : @0;
    NSNumber* timeout = arguments[@"time_out"] ? : @3;
    MPOS_RESULT nativeResult = [self.config openService:[deviceId integerValue] timeout:[timeout integerValue]];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)closeService:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* timeout = arguments[@"time_out"] ? : @3;
    MPOS_RESULT nativeResult = [self.config closeService:[timeout integerValue]];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)selectInterface:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* interfaceType = arguments[@"interface_type"] ? : @-1;
    NSString* address = arguments[@"address"] ? : @3;
    MPOS_RESULT nativeResult = [self.config selectInterface:[interfaceType integerValue] address:address];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)selectCommandMode:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* commandMode = arguments[@"command_mode"] ? : @-1;
    MPOS_RESULT nativeResult = [self.config selectCommandMode:[commandMode integerValue]];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)setTransaction:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* transaction = arguments[@"transaction"] ? : @-1;
    MPOS_RESULT nativeResult = [self.config setTransaction:[transaction integerValue]];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)setReadMode:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* mode = arguments[@"read_mode"] ? : @-1;
    MPOS_RESULT nativeResult = [self.config setReadMode:[mode integerValue]];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)directIO:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSArray<NSNumber*>* data = arguments[@"data_array"] ? : @[];
    
    NSMutableData* dataToSend = [NSMutableData data];
    for(NSNumber* number in data){
        char byte = [number charValue];
        [dataToSend appendBytes: &byte length:1];
    }
    MPOS_RESULT nativeResult = [self.config directIO:dataToSend];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)isOpen:(FlutterMethodCall*)call result:(FlutterResult)result {
    MPOS_RESULT nativeResult = [self.config isOpen] ? 1 : 0;
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)getDeviceId:(FlutterMethodCall*)call result:(FlutterResult)result {
    MPOS_RESULT nativeResult = [self.config getDeviceId];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)setKeepNetworkConnection:(FlutterMethodCall*)call result:(FlutterResult)result {
    if(![self.config isOpen]){
        NSDictionary* arguments = [call arguments];
        NSNumber* keep_connection = arguments[@"keep_connection"] ? : @0;
        [self.config setKeepNetworkConnection:[keep_connection integerValue]];
        result([NSNumber numberWithInteger: 0]);
    } else {
        result([NSNumber numberWithInteger: 1000]);
    }
}

- (void)isUsbPeripheralPrinter:(FlutterMethodCall*)call result:(FlutterResult)result {
    result([self.config isUSBPeripheralPrinter] ? @(YES) : @(NO));
}

// MARK: - Config APIs

- (void)searchDevices:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSArray<NSNumber*>* list = [self.config searchDevices];
    result([self convertJSONString:list]);
}

- (void)getBgateSerialNumber:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString* resultString = [self.config getBgateSerialNumber];
    result(resultString);
}

- (void)getUSBDevice:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* deviceId = arguments[@"device_id"] ? : @-1;
    NSString* resultString = [self.config getUSBDevice:[deviceId integerValue]];
    result(resultString);
}

- (void)getCustomDevices:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* device_type = arguments[@"device_type"] ? : @-1;
    NSArray<NSString*>* list = [self.config getCustomDevices:[device_type integerValue]];
    result(list);
}

- (void)getSerialConfig:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* device_id = arguments[@"device_id"] ? : @-1;
    NSArray<NSNumber*>* configInfo = [self.config getSerialConfig:[device_id integerValue]];
    result(configInfo);
}

- (void)getSerialConfiguration:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* device_id = arguments[@"device_id"] ? : @-1;
    NSArray<NSNumber*>* configInfo = [self.config getSerialConfiguration:[device_id integerValue]];
    result(configInfo);
}

- (void)setSerialConfig:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* device_id = arguments[@"device_id"] ? : @-1;
    NSNumber* baud_rate = arguments[@"baud_rate"] ? : @-1;
    NSNumber* data_bit = arguments[@"data_bit"] ? : @-1;
    NSNumber* stop_bit = arguments[@"stop_bit"] ? : @-1;
    NSNumber* parity_bit = arguments[@"parity_bit"] ? : @-1;
    
    MPOS_RESULT nativeResult = [self.config setSerialConfig:[device_id integerValue]
                                                   baudRate:[baud_rate integerValue]
                                                    dataBit:[data_bit integerValue]
                                                    stopBit:[stop_bit integerValue]
                                                  parityBit:[parity_bit integerValue]];
    
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)setSerialConfiguration:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* device_id = arguments[@"device_id"] ? : @-1;
    NSNumber* baud_rate = arguments[@"baud_rate"] ? : @2;
    NSNumber* data_bit = arguments[@"data_bit"] ? : @1;
    NSNumber* stop_bit = arguments[@"stop_bit"] ? : @0;
    NSNumber* parity_bit = arguments[@"parity_bit"] ? : @0;
    NSNumber* flow_control = arguments[@"flow_control"] ? : @0;
    
    MPOS_RESULT nativeResult = [self.config setSerialConfiguration:[device_id integerValue]
                                                          baudRate:[baud_rate integerValue]
                                                           dataBit:[data_bit integerValue]
                                                           stopBit:[stop_bit integerValue]
                                                         parityBit:[parity_bit integerValue]
                                                       flowControl:[flow_control integerValue]];
    
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)addCustomDevice:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* device_type = arguments[@"device_type"] ? : @-1;
    NSString* vid = arguments[@"vid"] ? : @"";
    NSString* pid = arguments[@"pid"] ? : @"";
    MPOS_RESULT nativeResult = [self.config addCustomDevice:[device_type integerValue] vid:vid pid:pid];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)deleteCustomDevice:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* device_type = arguments[@"device_type"] ? : @-1;
    NSString* vid = arguments[@"vid"] ? : @"";
    NSString* pid = arguments[@"pid"] ? : @"";
    MPOS_RESULT nativeResult = [self.config deleteCustomDevice:[device_type integerValue] vid:vid pid:pid];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (void)reInitCustomDeviceType:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary* arguments = [call arguments];
    NSNumber* device_type = arguments[@"device_type"] ? : @-1;
    MPOS_RESULT nativeResult = [self.config reInitCustomDeviceType:[device_type integerValue]];
    result([NSNumber numberWithInteger:nativeResult]);
}

- (NSInteger)getDeviceId {
    return [self.config getDeviceId];
}

- (NSString*) convertJSONString:(NSArray<NSNumber*>*)list {
    if (![list count]) {
        return @"[]";
    }
    
    NSMutableArray* jsonArray = [NSMutableArray array];
    
    for (NSNumber* deviceid in list) {
        NSMutableDictionary* jsonItem = [NSMutableDictionary dictionary];
        
        jsonItem[@"device_id"] = deviceid;
        
        if([deviceid integerValue] > 0 && [deviceid integerValue] < 10){
            jsonItem[@"device_type"] = @"LABEL PRINTER";
        }else if([deviceid integerValue] >= 10 && [deviceid integerValue] < 20){
            jsonItem[@"device_type"] = @"POS PRINTER";
        }else if([deviceid integerValue] >= 20 && [deviceid integerValue] < 30){
            jsonItem[@"device_type"] = @"Unregistered HID INPUT DEVICE";
        }else if([deviceid integerValue] >= 30 && [deviceid integerValue] < 40){
            jsonItem[@"device_type"] = @"MSR";
        }else if([deviceid integerValue] >= 40 && [deviceid integerValue] < 50){
            jsonItem[@"device_type"] = @"BARCODE SCANNER";
        }else if([deviceid integerValue] >= 60 && [deviceid integerValue] < 70){
            jsonItem[@"device_type"] = @"RFID";
        }else if([deviceid integerValue] >= 70 && [deviceid integerValue] < 80){
            jsonItem[@"device_type"] = @"DALLAS READER";
        }else if([deviceid integerValue] >= 80 && [deviceid integerValue] < 90){
            jsonItem[@"device_type"] = @"NFC";
        }else if([deviceid integerValue] >= 100 && [deviceid integerValue] < 110){
            jsonItem[@"device_type"] = @"Unregistered USB-SERIAL";
        }else if([deviceid integerValue] >= 110 && [deviceid integerValue] < 120){
            jsonItem[@"device_type"] = @"CUSTOMER DISPLAY";
        }else if([deviceid integerValue] >= 120 && [deviceid integerValue] < 130){
            jsonItem[@"device_type"] = @"USB-SERIAL";
        }else if([deviceid integerValue] >= 130 && [deviceid integerValue] < 140){
            jsonItem[@"device_type"] = @"USB-SCALE";
        }else{
            jsonItem[@"device_type"] = @"UNKNOWN";
        }
        [jsonArray addObject:jsonItem];
    }
    
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:jsonArray options:NSJSONWritingPrettyPrinted error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    
    return [NSString stringWithFormat:@"%@",jsonString];
}

@end
